tests/testsuite/registry_overlay.rs

```diff
@@ -0,0 +1,161 @@
+//! Tests for local-registry sources.
+
+use cargo_test_support::registry::{Package, RegistryBuilder, TestRegistry};
+use cargo_test_support::{project, Project};
+
+fn setup() -> (TestRegistry, String) {
+    let alt = RegistryBuilder::new().alternative().build();
+    (
+        RegistryBuilder::new().http_index().build(),
+        alt.index_url()
+            .to_file_path()
+            .unwrap()
+            .into_os_string()
+            .into_string()
+            .unwrap(),
+    )
+}
+
+fn simple_project() -> Project {
+    project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                edition = "2015"
+                authors = []
+
+                [dependencies]
+                bar = ">= 0.0.0"
+                baz = ">= 0.1.0"
+            "#,
+        )
+        .file("src/main.rs", "fn main() {}")
+        .build()
+}
+
+#[cargo_test]
+fn overlay_hit() {
+    let (reg, alt_path) = setup();
+    let p = simple_project();
+
+    Package::new("bar", "0.0.1").publish();
+    // baz is only in the local registry, but it gets found
+    Package::new("baz", "0.1.1")
+        .alternative(true)
+        .local(true)
+        .publish();
+
+    p.cargo("check")
+        .overlay_registry(&reg.index_url(), &alt_path)
+        .run();
+}
+
+#[cargo_test]
+fn version_fallback() {
+    let (reg, alt_path) = setup();
+    let p = simple_project();
+
+    Package::new("bar", "0.0.1").publish();
+    // The one we want is in the main registry.
+    Package::new("baz", "0.1.1").publish();
+    // An older version in the local registry doesn't confuse it.
+    Package::new("baz", "0.0.1")
+        .alternative(true)
+        .local(true)
+        .publish();
+
+    p.cargo("check")
+        .overlay_registry(&reg.index_url(), &alt_path)
+        .run();
+}
+
+#[cargo_test]
+fn version_collision() {
+    let (reg, alt_path) = setup();
+    let p = simple_project();
+
+    Package::new("bar", "0.0.1").publish();
+    // The one we want is in the main registry.
+    Package::new("baz", "0.1.1").publish();
+    Package::new("baz", "0.1.1")
+        .alternative(true)
+        .local(true)
+        .publish();
+
+    p.cargo("check")
+        .overlay_registry(&reg.index_url(), &alt_path)
+        .with_status(101)
+        .with_stderr(
+            "\
+[UPDATING] [..]
+[ERROR] failed to get `baz` [..]
+
+Caused by:
+  failed to query replaced source registry `crates-io`
+
+Caused by:
+  found a package in the remote registry and the local overlay: baz@0.1.1",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn multiversion() {
+    let (reg, alt_path) = setup();
+    let p = simple_project();
+
+    Package::new("bar", "0.0.1").publish();
+    Package::new("baz", "0.1.1").dep("baz", "=0.0.1").publish();
+    Package::new("baz", "0.0.1")
+        .alternative(true)
+        .local(true)
+        .publish();
```

Comment:
Should we have the local / selected version depend on the registry version?  Thats more like what we'll be dealing with for `cargo package`

Reply:
I don't think so, but also happy to discuss it further. My understanding was that the behavior of `cargo package` w.r.t. intra-workspace deps should be exactly the same as if the deps were published before packaging. Resolution should have the option to select the versions that are being packaged simultaneously, but it's also allowed not to.

Reply:
Yes, that should be the behavior but we aren't testing that situation to confirm it.  `cargo package` will involve a local package depending on an overlayed package which may depend on a registry package.

I guess there are also situations where you depend on a registry package which will depend on the overlaid package (if `clap_complete` used `criterion` it would both depend on clap directly and through `criterion`).  So maybe its best to cover both cases.

Reply:
Ah, makes sense, I just misunderstood your original comment. I'm now testing both directions.

> I guess there are also situations where you depend on a registry package which will depend on the overlaid package
I think the registry package will never *require* an overlaid package, right? Because the registry package can satisfy all its dependencies from the registry. I added a test where the registry package will pick up an overlaid dependency because it's newer (but compatible with) the registry version of that dependency.

Reply:
Not require but should resolve to use it.